### PR TITLE
Handle ObjC nullable annotations

### LINKED ISSUE #334: ObjC: Honor nullable annotations
[`sessionWithConfiguration:delegate:delegateQueue:`](https://developer.apple.com/documentation/foundation/nsurlsession/1411597-sessionwithconfiguration)  aka
```
+ (NSURLSession *)sessionWithConfiguration:(NSURLSessionConfiguration *)configuration delegate:(nullable id <NSURLSessionDelegate>)delegate delegateQueue:(nullable NSOperationQueue *)queue;
```

gets converted to:

```dart
  static NSURLSession sessionWithConfiguration_delegate_delegateQueue(
      NativeLibrary _lib,
      NSObject configuration,
      NSObject delegate,
      NSObject queue) {
```

it should be converted to:
```dart
  static NSURLSession sessionWithConfiguration_delegate_delegateQueue(
      NativeLibrary _lib,
      NSObject configuration,
      NSObject? delegate,
      NSObject? queue) {
```

### PR DESCRIPTION
**Pull request by liamappelbe**
Also, make ObjCNullable a proper type, rather than just a nullability flag on objc methods.

Fixes #334

Example (from nullable_test):

```Objective-C
@interface NullableInterface  : NSObject {
}

+(BOOL) isNullWithNullableNSObjectArg:(nullable NSObject *)x;
+(BOOL) isNullWithNotNullableNSObjectPtrArg:(NSObject *)x;
+(nullable NSObject *) returnNil:(BOOL)r;

@property (nullable, retain) NSObject *nullableObjectProperty;

@end
```

Generates:

```Dart
class NullableInterface extends NSObject {
  NullableInterface._(ffi.Pointer<ObjCObject> id, NullableTestObjCLibrary lib,
      {bool retain = false, bool release = false})
      : super._(id, lib, retain: retain, release: release);

  /// Returns a [NullableInterface] that points to the same underlying object as [other].
  static NullableInterface castFrom<T extends _ObjCWrapper>(T other) {
    return NullableInterface._(other._id, other._lib,
        retain: true, release: true);
  }

  /// Returns a [NullableInterface] that wraps the given raw object pointer.
  static NullableInterface castFromPointer(
      NullableTestObjCLibrary lib, ffi.Pointer<ObjCObject> other,
      {bool retain = false, bool release = false}) {
    return NullableInterface._(other, lib, retain: retain, release: release);
  }

  /// Returns whether [obj] is an instance of [NullableInterface].
  static bool isInstance(_ObjCWrapper obj) {
    return obj._lib._objc_msgSend_0(obj._id, obj._lib._sel_isKindOfClass_1,
        obj._lib._class_NullableInterface1);
  }

  static bool isNullWithNullableNSObjectArg_(
      NullableTestObjCLibrary _lib, NSObject? x) {
    return _lib._objc_msgSend_22(_lib._class_NullableInterface1,
        _lib._sel_isNullWithNullableNSObjectArg_1, x?._id ?? ffi.nullptr);
  }

  static bool isNullWithNotNullableNSObjectPtrArg_(
      NullableTestObjCLibrary _lib, NSObject x) {
    return _lib._objc_msgSend_23(_lib._class_NullableInterface1,
        _lib._sel_isNullWithNotNullableNSObjectPtrArg_1, x._id);
  }

  static NSObject? returnNil_(NullableTestObjCLibrary _lib, bool r) {
    final _ret = _lib._objc_msgSend_24(
        _lib._class_NullableInterface1, _lib._sel_returnNil_1, r);
    return _ret.address == 0
        ? null
        : NSObject._(_ret, _lib, retain: true, release: true);
  }

  NSObject? get nullableObjectProperty {
    final _ret = _lib._objc_msgSend_25(_id, _lib._sel_nullableObjectProperty1);
    return _ret.address == 0
        ? null
        : NSObject._(_ret, _lib, retain: true, release: true);
  }

  set nullableObjectProperty(NSObject? value) {
    return _lib._objc_msgSend_26(
        _id, _lib._sel_setNullableObjectProperty_1, value?._id ?? ffi.nullptr);
  }

  @override
  NullableInterface init() {
    final _ret = _lib._objc_msgSend_2(_id, _lib._sel_init1);
    return NullableInterface._(_ret, _lib, retain: true, release: true);
  }

  static NullableInterface new1(NullableTestObjCLibrary _lib) {
    final _ret =
        _lib._objc_msgSend_2(_lib._class_NullableInterface1, _lib._sel_new1);
    return NullableInterface._(_ret, _lib, retain: false, release: true);
  }

  static NullableInterface allocWithZone_(
      NullableTestObjCLibrary _lib, ffi.Pointer<_NSZone> zone) {
    final _ret = _lib._objc_msgSend_3(
        _lib._class_NullableInterface1, _lib._sel_allocWithZone_1, zone);
    return NullableInterface._(_ret, _lib, retain: false, release: true);
  }

  static NullableInterface alloc(NullableTestObjCLibrary _lib) {
    final _ret =
        _lib._objc_msgSend_2(_lib._class_NullableInterface1, _lib._sel_alloc1);
    return NullableInterface._(_ret, _lib, retain: false, release: true);
  }
}
```

Diff:

```diff
32,34c32,34
<       NullableTestObjCLibrary _lib, NSObject? x) {
<     return _lib._objc_msgSend_22(_lib._class_NullableInterface1,
<         _lib._sel_isNullWithNotNullableNSObjectPtrArg_1, x?._id ?? ffi.nullptr);
---
>       NullableTestObjCLibrary _lib, NSObject x) {
>     return _lib._objc_msgSend_23(_lib._class_NullableInterface1,
>         _lib._sel_isNullWithNotNullableNSObjectPtrArg_1, x._id);
37,38c37,38
<   static NSObject returnNil_(NullableTestObjCLibrary _lib, bool r) {
---
>   static NSObject? returnNil_(NullableTestObjCLibrary _lib, bool r) {
40c40,42
<     return NSObject._(_ret, _lib, retain: true, release: true);
---
>     return _ret.address == 0
>         ? null
>         : NSObject._(_ret, _lib, retain: true, release: true);
```

### CODE REVIEW

**Comment by dcharkes** [comment]
The CI is still red, is this ready for review or WIP?

**Comment by liamappelbe** [comment]
> The CI is still red, is this ready for review or WIP?

Ready for review. The fixes to these tests should be pretty minor. I'm working on them now.

File: lib/src/code_generator/objc_nullable.dart [hunk 1]
```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2023, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:ffigen/src/code_generator.dart';
+
+import 'writer.dart';
+
+/// An ObjC type annotated with nullable.
```
Comment (dcharkes) [inline lib/src/code_generator/objc_nullable.dart:9]: Nit: Add ObjC code snippet as example.

File: lib/src/code_generator/objc_nullable.dart [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+class ObjCNullable extends Type {
```
Comment (dcharkes) [inline lib/src/code_generator/objc_nullable.dart:10]: Nice, much cleaner!

**Review by dcharkes — APPROVED** [review]
lgtm

nit: now that we don't check in generated files anymore, please add a pre-and-post FFIgen output of a small example to the PR description so it's easy to see the diff.

**Pull request merged**